File: Packages/Flip54WorkoutEngine/Sources/Flip54WorkoutEngine/WorkoutCoordinator.swift
import Foundation
import Flip54Core
import Flip54Storage

/// Owns the workout state machine. All mutations go through `send(_:)`.
/// Persists the active session on every state change.
@MainActor
@Observable
public final class WorkoutCoordinator {
    public private(set) var state: WorkoutState = .idle
    public private(set) var session: ActiveSession?
    /// True when running a tutorial flip — coordinator won't persist session and caller skips history save.
    public private(set) var isTutorial: Bool = false

    private let store: ActiveSessionStore
    private let holdTimer: HoldTimer

    public init(store: ActiveSessionStore = ActiveSessionStore()) {
        self.store = store
        self.holdTimer = HoldTimer()
        self.holdTimer.onExpired = { [weak self] in
            Task { @MainActor [weak self] in
                self?.send(.holdTimerExpired)
            }
        }
    }

    // MARK: - Public API

    public func send(_ event: WorkoutEvent) {
        let next = transition(state: state, event: event)
        guard next != state else { return }
        applyTransition(from: state, to: next, event: event)
        state = next
    }

    // MARK: - Transition table

    private func transition(state: WorkoutState, event: WorkoutEvent) -> WorkoutState {
        switch (state, event) {

        case (.idle, .shuffle):
            return .shuffling

        case (.shuffling, .shuffleComplete):
            return .cardFaceDown

        case (.cardFaceDown, .flipCard):
            guard var s = session, !s.drawPile.isEmpty else { return state }
            s.flipNextCard()
            session = s
            guard let card = s.currentCard else { return state }
            let p = prescription(for: card, equipment: s.equipment, difficulty: s.difficulty)
            return .cardFaceUp(card: card, prescription: p)

        case (.cardFaceUp(let card, _), .startHold) where card.isAce:
            return .holdStarting(card: card)

        case (.holdStarting(let card), .startHold):
            guard let s = session else { return state }
            let p = prescription(for: card, equipment: s.equipment, difficulty: s.difficulty)
            guard case .hold(_, let secs) = p else { return state }
            return .holding(card: card, startTime: Date(), durationSeconds: secs)

        case (.cardFaceUp(let card, _), .markDone(let reps, let secs)):
            return .cardCompleting(card: card)

        case (.holding(let card, _, _), .markDone(_, let holdSecs)):
            let held = holdSecs ?? 0
            return .holdComplete(card: card, secondsHeld: held, completedFully: false)

        case (.holding(let card, let start, let dur), .holdTimerExpired):
            let held = Int(Date().timeIntervalSince(start))
            return .holdComplete(card: card, secondsHeld: min(held, dur), completedFully: true)

        case (.holdComplete(let card, _, _), .advanceComplete):
            return .cardCompleting(card: card)

        case (.cardFaceUp(let card, _), .skip),
             (.holdStarting(let card), .skip),
             (.holding(let card, _, _), .skip):
            return .cardSkipping(card: card)

        case (.cardCompleting, .advanceComplete):
            guard let s = session else { return .workoutComplete }
            return s.isComplete ? .workoutComplete : .cardFaceDown

        case (.cardSkipping, .advanceComplete):
            return .cardFaceDown

        case (_, .pause) where !isPaused(state):
            return .paused(previous: state)

        case (.paused(let prev), .resume):
            return prev

        default:
            return state
        }
    }

    // MARK: - Side effects

    private func applyTransition(from: WorkoutState, to: WorkoutState, event: WorkoutEvent) {
        switch to {
        case .shuffling:
            startNewSession()

        case .holding(_, let start, let dur):
            var elapsed: TimeInterval = 0
            if let s = session { elapsed = s.holdElapsed }
            holdTimer.start(durationSeconds: dur, alreadyElapsed: elapsed)
            session?.startHold(at: start)

        case .holdComplete(let card, let secs, _):
            holdTimer.cancel()
            if card.isAce {
                session?.completeCurrentCard(reps: nil, holdSeconds: secs)
            }

        case .cardCompleting(let card):
            if case .cardFaceUp(_, let p) = from, case .reps(_, let count) = p {
                session?.completeCurrentCard(reps: count, holdSeconds: nil)
            }
            persistSession()

        case .cardSkipping:
            holdTimer.cancel()
            session?.skipCurrentCard()
            persistSession()

        case .workoutComplete:
            store.clear()

        case .paused(let prev):
            session?.pause(at: Date())
            holdTimer.cancel()
            persistSession()

        case .cardFaceDown:
            if case .paused(_) = from {
                session?.resume(at: Date())
                if case .holding(_, let start, let dur) = from {
                    let elapsed = session?.holdElapsed ?? 0
                    holdTimer.start(durationSeconds: dur, alreadyElapsed: elapsed)
                }
            }

        default:
            break
        }
    }

    // MARK: - Helpers

    private func startNewSession() {
        isTutorial = false
        var rng = SystemRandomNumberGenerator()
        let equipment = session?.equipment ?? .bodyWeightOnly
        let difficulty = session?.difficulty ?? .standard
        session = ActiveSession.start(
            deck: Card.standardDeck(),
            deckId: session?.deckId ?? "standard",
            equipment: equipment,
            difficulty: difficulty,
            rng: &rng
        )
        persistSession()
    }

    private func persistSession() {
        guard !isTutorial, let s = session else { return }
        try? store.save(s)
    }

    private func isPaused(_ state: WorkoutState) -> Bool {
        if case .paused = state { return true }
        return false
    }

    // MARK: - Resume support

    public func restoreIfNeeded(equipment: Equipment, difficulty: Difficulty, deckId: String) {
        if let saved = store.load(), !saved.isComplete,
           Date().timeIntervalSince(saved.startedAt) < 86400 {
            session = saved
            state = .cardFaceDown
        } else {
            store.clear()
            session = nil
        }
    }

    public func configure(equipment: Equipment, difficulty: Difficulty, deckId: String) {
        var rng = SystemRandomNumberGenerator()
        session = ActiveSession.start(
            deck: Card.standardDeck(),
            deckId: deckId,
            equipment: equipment,
            difficulty: difficulty,
            rng: &rng
        )
    }

    /// Configure a 5-card tutorial deck: regular → ace → face → joker → regular (for skip demo).
    /// Results are NOT saved to history; caller checks `isTutorial` before persisting.
    public func configureTutorial(equipment: Equipment, difficulty: Difficulty) {
        isTutorial = true
        let tutorialCards: [Card] = [
            .standard(suit: .hearts,  rank: .seven),   // page 1: basic reps
            .standard(suit: .spades,  rank: .ace),     // page 2: hold exercise
            .standard(suit: .clubs,   rank: .king),    // page 3: face card
            .joker(variant: .red),                     // page 4: joker / jumping jacks
            .standard(suit: .diamonds, rank: .five),   // page 5: regular + hint to skip
        ]
        var rng = SystemRandomNumberGenerator()
        session = ActiveSession.start(
            deck: tutorialCards,
            deckId: "tutorial",
            equipment: equipment,
            difficulty: .beginner,  // reduced reps for tutorial
            rng: &rng
        )
        // Tutorial deck is already in the desired order — override the shuffled pile
        session?.drawPile = tutorialCards
        session?.currentCard = nil
    }

    /// Seed a specific draw pile for testing purposes.
    public func _seedDrawPile(_ cards: [Card]) {
        session?.drawPile = cards
        session?.currentCard = nil
    }
}

File: android/engine/src/main/kotlin/com/flip54/engine/WorkoutCoordinator.kt
package com.flip54.engine

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.flip54.core.*
import com.flip54.storage.ActiveSession
import com.flip54.storage.ActiveSessionStore
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import javax.inject.Inject

@HiltViewModel
class WorkoutCoordinator @Inject constructor(
    private val store: ActiveSessionStore
) : ViewModel() {

    private val _state = MutableStateFlow<WorkoutState>(WorkoutState.Idle)
    val state: StateFlow<WorkoutState> = _state.asStateFlow()

    private val _session = MutableStateFlow<ActiveSession?>(null)
    val session: StateFlow<ActiveSession?> = _session.asStateFlow()

    private val _isTutorial = MutableStateFlow(false)
    val isTutorial: StateFlow<Boolean> = _isTutorial.asStateFlow()

    private var useHalfDeck: Boolean = false

    private val holdTimer = HoldTimer(viewModelScope).also { timer ->
        timer.onExpired = { send(WorkoutEvent.HoldTimerExpired) }
    }

    val holdTimerSeconds: StateFlow<Int> = holdTimer.remainingSeconds

    // MARK: - Public API

    fun send(event: WorkoutEvent) {
        val current = _state.value
        var next = transition(current, event)
        if (next == current) return

        // On resume from pause into holding: rebuild startTime accounting for elapsed hold
        if (current is WorkoutState.Paused && next is WorkoutState.Holding) {
            val elapsedMs = _session.value?.holdElapsedMs ?: 0L
            next = WorkoutState.Holding(
                card = next.card,
                startTimeMs = System.currentTimeMillis() - elapsedMs,
                durationSeconds = next.durationSeconds
            )
        }

        applyTransition(from = current, to = next, event = event)
        _state.value = next
    }

    // MARK: - State machine

    private fun transition(state: WorkoutState, event: WorkoutEvent): WorkoutState {
        return when {
            state is WorkoutState.Idle && event is WorkoutEvent.Shuffle ->
                WorkoutState.Shuffling

            state is WorkoutState.Shuffling && event is WorkoutEvent.ShuffleComplete ->
                WorkoutState.CardFaceDown

            state is WorkoutState.CardFaceDown && event is WorkoutEvent.FlipCard -> {
                val s = _session.value ?: return state
                if (s.drawPile.isEmpty()) return state
                val flipped = s.flipNextCard()
                _session.value = flipped
                val card = flipped.currentCard ?: return state
                val p = prescription(card, s.equipment, s.difficulty)
                WorkoutState.CardFaceUp(card, p)
            }

            state is WorkoutState.CardFaceUp && state.card.isAce && event is WorkoutEvent.StartHold ->
                WorkoutState.HoldStarting(state.card)

            state is WorkoutState.HoldStarting && event is WorkoutEvent.StartHold -> {
                val s = _session.value ?: return state
                val p = prescription(state.card, s.equipment, s.difficulty)
                val secs = (p as? Prescription.Hold)?.seconds ?: return state
                WorkoutState.Holding(state.card, System.currentTimeMillis(), secs)
            }

            state is WorkoutState.CardFaceUp && event is WorkoutEvent.MarkDone ->
                WorkoutState.CardCompleting(state.card)

            state is WorkoutState.Holding && event is WorkoutEvent.MarkDone ->
                WorkoutState.HoldComplete(state.card, event.holdSeconds ?: 0, false)

            state is WorkoutState.Holding && event is WorkoutEvent.HoldTimerExpired -> {
                val held = ((System.currentTimeMillis() - state.startTimeMs) / 1000).toInt()
                WorkoutState.HoldComplete(state.card, minOf(held, state.durationSeconds), true)
            }

            state is WorkoutState.HoldComplete && event is WorkoutEvent.AdvanceComplete ->
                WorkoutState.CardCompleting(state.card)

            (state is WorkoutState.CardFaceUp || state is WorkoutState.HoldStarting ||
                    state is WorkoutState.Holding) && event is WorkoutEvent.Skip -> {
                val card = when (state) {
                    is WorkoutState.CardFaceUp   -> state.card
                    is WorkoutState.HoldStarting -> state.card
                    is WorkoutState.Holding      -> state.card
                    else -> return state
                }
                WorkoutState.CardSkipping(card)
            }

            state is WorkoutState.CardCompleting && event is WorkoutEvent.AdvanceComplete -> {
                val s = _session.value ?: return WorkoutState.WorkoutComplete
                if (s.isComplete) WorkoutState.WorkoutComplete else WorkoutState.CardFaceDown
            }

            state is WorkoutState.CardSkipping && event is WorkoutEvent.AdvanceComplete ->
                WorkoutState.CardFaceDown

            !state.isPaused && event is WorkoutEvent.Pause ->
                WorkoutState.Paused(state)

            state is WorkoutState.Paused && event is WorkoutEvent.Resume ->
                state.previous

            else -> state
        }
    }

    // MARK: - Side effects

    private fun applyTransition(from: WorkoutState, to: WorkoutState, event: WorkoutEvent) {
        when (to) {
            is WorkoutState.Shuffling -> startNewSession()

            is WorkoutState.Holding -> {
                val elapsedMs = _session.value?.holdElapsedMs ?: 0L
                holdTimer.start(to.durationSeconds, elapsedMs)
                if (from is WorkoutState.Paused) {
                    _session.value = _session.value?.resume(System.currentTimeMillis())
                }
                _session.value = _session.value?.startHold(System.currentTimeMillis())
            }

            is WorkoutState.HoldComplete -> {
                holdTimer.cancel()
                if (to.card.isAce) {
                    _session.value = _session.value?.completeCurrentCard(
                        reps = null,
                        holdSeconds = to.secondsHeld,
                        nowMs = System.currentTimeMillis()
                    )
                }
            }

            is WorkoutState.CardCompleting -> {
                if (from is WorkoutState.CardFaceUp) {
                    val count = (from.prescription as? Prescription.Reps)?.count
                    if (count != null) {
                        _session.value = _session.value?.completeCurrentCard(
                            reps = count,
                            holdSeconds = null,
                            nowMs = System.currentTimeMillis()
                        )
                    }
                }
                persistSession()
            }

            is WorkoutState.CardSkipping -> {
                holdTimer.cancel()
                _session.value = _session.value?.skipCurrentCard()
                persistSession()
            }

            is WorkoutState.WorkoutComplete -> store.clear()

            is WorkoutState.Paused -> {
                _session.value = _session.value?.pause(System.currentTimeMillis())
                holdTimer.cancel()
                persistSession()
            }

            is WorkoutState.CardFaceDown -> {
                if (from is WorkoutState.Paused) {
                    _session.value = _session.value?.resume(System.currentTimeMillis())
                }
            }

            else -> Unit
        }
    }

    // MARK: - Session management

    private fun startNewSession() {
        if (_isTutorial.value) return
        val current = _session.value
        val equipment = current?.equipment ?: Equipment.bodyWeightOnly
        val difficulty = current?.difficulty ?: Difficulty.STANDARD
        val deckId = current?.deckId ?: "standard"
        val deck = if (useHalfDeck) Card.halfDeck() else Card.standardDeck()
        val session = ActiveSession.start(deck, deckId, equipment, difficulty)
        _session.value = session
        persistSession()
    }

    private fun persistSession() {
        if (_isTutorial.value) return
        _session.value?.let { store.save(it) }
    }

    // MARK: - Configuration

    fun configure(equipment: Equipment, difficulty: Difficulty, deckId: String, useHalfDeck: Boolean = false) {
        _isTutorial.value = false
        this.useHalfDeck = useHalfDeck
        val deck = if (useHalfDeck) Card.halfDeck() else Card.standardDeck()
        _session.value = ActiveSession.start(deck, deckId, equipment, difficulty)
    }

    fun configureTutorial(equipment: Equipment, difficulty: Difficulty) {
        _isTutorial.value = true
        val tutorialCards = listOf(
            Card.Standard(Suit.HEARTS,   Rank.SEVEN),
            Card.Standard(Suit.SPADES,   Rank.KING),
            Card.Standard(Suit.CLUBS,    Rank.ACE),
            Card.Standard(Suit.DIAMONDS, Rank.FIVE),
            Card.Joker(JokerVariant.RED)
        )
        val session = ActiveSession.start(tutorialCards, "tutorial", equipment, Difficulty.BEGINNER)
        _session.value = session.copy(drawPile = tutorialCards, currentCard = null)
    }

    fun restoreIfNeeded(equipment: Equipment, difficulty: Difficulty, deckId: String) {
        val saved = store.load()
        val nowMs = System.currentTimeMillis()
        if (saved != null && !saved.isComplete && (nowMs - saved.startedAtMs) < 86_400_000L) {
            _isTutorial.value = false
            _session.value = saved
            _state.value = WorkoutState.CardFaceDown
        } else {
            store.clear()
            _session.value = null
        }
    }

    fun endEarly() {
        holdTimer.cancel()
        if (!_isTutorial.value) {
            _session.value?.let { store.save(it) }
        }
        _session.value = null
        _isTutorial.value = false
        _state.value = WorkoutState.Idle
    }

    fun reset() {
        holdTimer.cancel()
        store.clear()
        _session.value = null
        _isTutorial.value = false
        _state.value = WorkoutState.Idle
    }
}
